### DISCUSSION TITLE
Default state in model attributes overrides state in DB

### DISCUSSION THREAD

**Discussion by ragulka** [spatie/laravel-model-states#232] [Bugs]
Let's say I have the following model, which defines a default attribute value for the 'state' property:

```php
class Payment extends Model {
    
    protected $attributes = [
        'state' => 'pending'
    ];
    
    protected $casts = [
        'state' => PaymentState::class
    ];
}
```

When I have a payment with ID 1 in my DB where state is `paid`, and I call `Payment::find(1)->status`, it will return `PendingState::class` instead of `PaidState::class`.

Interestingly:
* If I remove the default `state` from `protected $attributes`, this issue does not occur.
* If I remove `state` from `protected $casts` (and keep the default in `$attributes`), this issue does not occur.

I would expect it to return `PaidState::class`.

I did quiet a bit of digging, and it seems this happens because after the model is loaded and attributes from DB set on the model, `syncOriginal` is called and [it will merge attributes from cached casts](https://github.com/laravel/framework/blob/4b3bd091d7a4e8a9c205cf528d3f08c345c8c1be/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L1714-L1726). This will call the class caster set method, but not with the raw value, but the default attribute value (which will be an instance of `PendingState` in our case).

I'm not sure why it works this way, but it seems like a bug - although probably on Laravel side. I wanted to raise this concern regardless, as it threw me off quiet a bit.

A workaround for the issue is to not define the default state value in model attributes, but rather in the [state class](https://spatie.be/docs/laravel-model-states/v2/working-with-states/01-configuring-states).

**Comment by ragulka** [comment]
I dug a little deeper, as I tried to reproduce this with Laravel using another class cast. At first, I could not reproduce the issue. Then I realized the `HasStates` trait has an initalizer, which calls `setDefaultState`, which [checks the default attribute value here](https://github.com/spatie/laravel-model-states/blob/main/src/HasStates.php#L143-L145).

This  seems to be the culprit - accessing the model value this way calls the magic getter, which calls the caster with the default value (not the DB value, because those have not been set from the DB, yet), and then caches the result in class cast. Then, when `syncOriginal` is called, the DB value is ignored, because the class cast cache already contains a value for the attribute.

Basically, it can be said that accessing the model property has an undesirable side-effect here. I'm kind of torn wether this is a Laravel bug, or a bug in this package.

In any case, it seems that modifying [this line here](https://github.com/spatie/laravel-model-states/blob/main/src/HasStates.php#L143) as shown, fixes (works around) the issue:

```php
if (Arr::get($this->attributes, $field) !== null) 
```
